lib/segment/src/index/struct_payload_index.rs

```diff
@@ -483,18 +482,18 @@ impl PayloadIndex for StructPayloadIndex {
         field: PayloadKeyTypeRef,
         payload_schema: &PayloadFieldSchema,
         hw_counter: &HardwareCounterCell,
-    ) -> OperationResult<Option<Vec<FieldIndex>>> {
+    ) -> OperationResult<BuildIndexResult> {
         if let Some(prev_schema) = self.config.indexed_fields.get(field) {
             // the field is already indexed with the same schema
             // no need to rebuild index and to save the config
-            if prev_schema == payload_schema {
-                return Ok(None);
-            }
+            return if prev_schema == payload_schema {
+                Ok(BuildIndexResult::AlreadyBuilt)
+            } else {
+                Ok(BuildIndexResult::IncompatibleSchema)
```

Comment:
We can't handle `IncompatibleSchema` inside `build_index` function, cause this function only has read-only access to the segment. This is important, since we want to split index building and index registration. Index building is a long process, and we don't want to build it while holding write lock. Therefore we have to process `IncompatibleSchema` externally